<a href="https://colab.research.google.com/github/ChimSovann/Machine-Learning/blob/main/KNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#lib
import sklearn
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn import preprocessing

In [24]:
#dataset
from google.colab import files
upload = files.upload ()

Saving Iris.csv to Iris (1).csv


In [25]:
#dataset
iris = pd.read_csv ('Iris.csv')
iris.head(5)

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

In [26]:
iris.shape

(150, 6)

In [27]:
iris.columns

Index(['Id', 'SepalLengthCm', 'SepalWidthCm', 'PetalLengthCm', 'PetalWidthCm',
       'Species'],
      dtype='object')

In [28]:
iris.values

array([[1, 5.1, 3.5, 1.4, 0.2, 'Iris-setosa'],
       [2, 4.9, 3.0, 1.4, 0.2, 'Iris-setosa'],
       [3, 4.7, 3.2, 1.3, 0.2, 'Iris-setosa'],
       [4, 4.6, 3.1, 1.5, 0.2, 'Iris-setosa'],
       [5, 5.0, 3.6, 1.4, 0.2, 'Iris-setosa'],
       [6, 5.4, 3.9, 1.7, 0.4, 'Iris-setosa'],
       [7, 4.6, 3.4, 1.4, 0.3, 'Iris-setosa'],
       [8, 5.0, 3.4, 1.5, 0.2, 'Iris-setosa'],
       [9, 4.4, 2.9, 1.4, 0.2, 'Iris-setosa'],
       [10, 4.9, 3.1, 1.5, 0.1, 'Iris-setosa'],
       [11, 5.4, 3.7, 1.5, 0.2, 'Iris-setosa'],
       [12, 4.8, 3.4, 1.6, 0.2, 'Iris-setosa'],
       [13, 4.8, 3.0, 1.4, 0.1, 'Iris-setosa'],
       [14, 4.3, 3.0, 1.1, 0.1, 'Iris-setosa'],
       [15, 5.8, 4.0, 1.2, 0.2, 'Iris-setosa'],
       [16, 5.7, 4.4, 1.5, 0.4, 'Iris-setosa'],
       [17, 5.4, 3.9, 1.3, 0.4, 'Iris-setosa'],
       [18, 5.1, 3.5, 1.4, 0.3, 'Iris-setosa'],
       [19, 5.7, 3.8, 1.7, 0.3, 'Iris-setosa'],
       [20, 5.1, 3.8, 1.5, 0.3, 'Iris-setosa'],
       [21, 5.4, 3.4, 1.7, 0.2, 'Iris-setosa'],
 

In [29]:
x_variable = ['SepalLengthCm', 'SepalWidthCm', 'PetalLengthCm', 'PetalWidthCm'] #input, Multiple input
y_variable = ['Species'] #output

#reshape variable
x = iris[x_variable].values.reshape(-1,4) # 1st value is -1, next value is the length of x_variable
y = iris[y_variable].values.reshape(-1,1)
print(x.shape)
print(y.shape)

(150, 4)
(150, 1)


In [30]:
#split the dataset
x_train, x_test,  y_train, y_test = train_test_split(x, y , test_size = 0.2, random_state=1)
print(x_train.shape)
print(x_test.shape)

(120, 4)
(30, 4)


In [31]:
from sklearn.neighbors import KNeighborsClassifier
knn_model = KNeighborsClassifier (n_neighbors=5,n_jobs=-1).fit(x_train,y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


In [32]:
x_test.shape

(30, 4)

In [33]:
y_predict = knn_model.predict (x_test)
print(y_predict)

['Iris-setosa' 'Iris-versicolor' 'Iris-versicolor' 'Iris-setosa'
 'Iris-virginica' 'Iris-versicolor' 'Iris-virginica' 'Iris-setosa'
 'Iris-setosa' 'Iris-virginica' 'Iris-versicolor' 'Iris-setosa'
 'Iris-virginica' 'Iris-versicolor' 'Iris-versicolor' 'Iris-setosa'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-setosa' 'Iris-setosa'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor' 'Iris-setosa'
 'Iris-virginica' 'Iris-versicolor' 'Iris-setosa' 'Iris-setosa'
 'Iris-versicolor' 'Iris-virginica']


In [ ]:
#dataframe of predict and actual values

In [51]:
#1. Create a table between actual and predict
from tabulate import tabulate
li = y_test.reshape(-1)
print(li)
l = list(zip(li, y_predict))
lists = [list(x) for x in l]
pred_y_df = pd.DataFrame({'Actual Value' : lists,'Predicted Value' : y_predict})
pred_y_df[0:25]

['Iris-setosa' 'Iris-versicolor' 'Iris-versicolor' 'Iris-setosa'
 'Iris-virginica' 'Iris-versicolor' 'Iris-virginica' 'Iris-setosa'
 'Iris-setosa' 'Iris-virginica' 'Iris-versicolor' 'Iris-setosa'
 'Iris-virginica' 'Iris-versicolor' 'Iris-versicolor' 'Iris-setosa'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-setosa' 'Iris-setosa'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor' 'Iris-setosa'
 'Iris-virginica' 'Iris-versicolor' 'Iris-setosa' 'Iris-setosa'
 'Iris-versicolor' 'Iris-virginica']


Actual Value  Predicted Value
0           [Iris-setosa, Iris-setosa]      Iris-setosa
1   [Iris-versicolor, Iris-versicolor]  Iris-versicolor
2   [Iris-versicolor, Iris-versicolor]  Iris-versicolor
3           [Iris-setosa, Iris-setosa]      Iris-setosa
4     [Iris-virginica, Iris-virginica]   Iris-virginica
5   [Iris-versicolor, Iris-versicolor]  Iris-versicolor
6     [Iris-virginica, Iris-virginica]   Iris-virginica
7           [Iris-setosa, Iris-setosa]      Iris-setosa
8           [Iris-setosa, Iris-setosa]      Iris-setosa
9     [Iris-virginica, Iris-virginica]   Iris-virginica
10  [Iris-versicolor, Iris-versicolor]  Iris-versicolor
11          [Iris-setosa, Iris-setosa]      Iris-setosa
12    [Iris-virginica, Iris-virginica]   Iris-virginica
13  [Iris-versicolor, Iris-versicolor]  Iris-versicolor
14  [Iris-versicolor, Iris-versicolor]  Iris-versicolor
15          [Iris-setosa, Iris-setosa]      Iris-setosa
16  [Iris-versicolor, Iris-versicolor]  Iris-versicolor
17  [Iris-versicolor, Iris-versicolor]  Iris-versicolor
18          [Iris-setosa, Iris-setosa]      Iris-setosa
19          [Iris-setosa, Iris-setosa]      Iris-setosa
20  [Iris-versicolor, Iris-versicolor]  Iris-versicolor
21  [Iris-versicolor, Iris-versicolor]  Iris-versicolor
22  [Iris-versicolor, Iris-versicolor]  Iris-versicolor
23          [Iris-setosa, Iris-setosa]      Iris-setosa
24    [Iris-virginica, Iris-virginica]   Iris-virginica

In [ ]:
knn_model.predict_proba (x_test)

array([[1. , 0. , 0. ],
       [0. , 1. , 0. ],
       [0. , 1. , 0. ],
       [1. , 0. , 0. ],
       [0. , 0. , 1. ],
       [0. , 0.8, 0.2],
       [0. , 0. , 1. ],
       [1. , 0. , 0. ],
       [1. , 0. , 0. ],
       [0. , 0. , 1. ],
       [0. , 1. , 0. ],
       [1. , 0. , 0. ],
       [0. , 0. , 1. ],
       [0. , 1. , 0. ],
       [0. , 1. , 0. ],
       [1. , 0. , 0. ],
       [0. , 1. , 0. ],
       [0. , 1. , 0. ],
       [1. , 0. , 0. ],
       [1. , 0. , 0. ],
       [0. , 1. , 0. ],
       [0. , 1. , 0. ],
       [0. , 0.6, 0.4],
       [1. , 0. , 0. ],
       [0. , 0. , 1. ],
       [0. , 1. , 0. ],
       [1. , 0. , 0. ],
       [1. , 0. , 0. ],
       [0. , 1. , 0. ],
       [0. , 0.4, 0.6]])

In [ ]:
input = np.array([7.7,2.6,6.9,2.3])
input = input.reshape(-1,4)

In [ ]:
y_predict1 = knn_model.predict(input)
print(y_predict)

['Iris-setosa' 'Iris-versicolor' 'Iris-versicolor' 'Iris-setosa'
 'Iris-virginica' 'Iris-versicolor' 'Iris-virginica' 'Iris-setosa'
 'Iris-setosa' 'Iris-virginica' 'Iris-versicolor' 'Iris-setosa'
 'Iris-virginica' 'Iris-versicolor' 'Iris-versicolor' 'Iris-setosa'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-setosa' 'Iris-setosa'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor' 'Iris-setosa'
 'Iris-virginica' 'Iris-versicolor' 'Iris-setosa' 'Iris-setosa'
 'Iris-versicolor' 'Iris-virginica']


In [ ]:
knn_model.predict_proba(input)

array([[0., 0., 1.]])

In [ ]:
#accuracy of the test dataset
from sklearn.metrics import accuracy_score
accuracy_result = accuracy_score (y_test, y_predict)
print (accuracy_result)

1.0



1.  Lib
1.  Dataset
2.  Preprocessing 3.1.X_variables, y_variables 3.2.Reshape
1.  Model build
1. Predict the result (whole dataset, single value)
2. Probability
2. Evaluation metrics(Accuracy) Accuracy score(single)











Task


1. Create a table between actual and predict
2. Check the formula of accuracy score
1.find the procedure to get the optimal k-value







# 2. Check the formular of accuracy score
#Accuracy = Number of Correct Predictions / Total Number of Prediction

#3.find the procedure to get the optimal k-value
1.   There are no pre-defined statistical methods to find the most favorable value of K.
2.   Initialize a random K value and start computing.
1.   Choosing a small value of K leads to unstable decision boundaries.
2.   The substantial K value is better for classification as it leads to smoothening the decision boundaries.
1.   Derive a plot between error rate and K denoting values in a defined range.
2.   hen choose the K value as having a minimum error rate.
